### PR TITLE
fix: multiple-notifs-when-ongoing

### PR DESCRIPTION
**Pull request by zzjjaayy**
## Issue
The app right now doesn't keep track of the notification id thus creating multiple notifications in case of ongoing notifications like download progress.

## Fix
- Switched the primary key to keep track of the notificationId + packageName (composite primary key) instead of auto generating. {this helps in keeping the same notification object with updated progress}
- Added Migration for db version bump from (1 to 2)

## Video Recording 
The start of the video shows a new notification being created at each change. The restart of the app is the install of the update and shows the progress update in the same notification.

https://github.com/darshanpania/Notificity/assets/54902986/5a44a60e-aa55-4656-a0b3-cf546037f7b6

### CODE REVIEW

**Comment by darshanpania** [comment]
@zzjjaayy I have 2 PRs on the UI bit that I need to first check and then I will come to this. Putting this on hold for a day or two. The problem is crucial for a fix, so I will circle back once I sort out the UI. 🙏

**Comment by darshanpania** [comment]
@zzjjaayy Pushed some changes to the main branch. Please rebase with the new changes

File: app/src/main/java/com/darshan/notificity/Notification.kt [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
 
-import android.graphics.Bitmap
 import androidx.room.Entity
-import androidx.room.PrimaryKey
 
-@Entity
-data class NotificationEntity(
-    @PrimaryKey(autoGenerate = true) val id: Int = 0,
+@Entity(tableName = "notification", primaryKeys = ["id", "packageName"])
+data class Notification(
```
Comment (darshanpania) [inline app/src/main/java/com/darshan/notificity/Notification.kt:6]: @zzjjaayy Can you let the class name be the same? That way we save a lot of changes

File: app/src/main/java/com/darshan/notificity/AppDatabase.kt [hunk 2]
```diff
@@ -25,13 +26,40 @@ abstract class AppDatabase : RoomDatabase() {
...
+
+        /**
+        * MIGRATION 1 -> 2
+        * Change in PrimaryKey for [Notification]
+         * The primary key is now composed of [Notification.id] && [Notification.packageName]
+         * The [Notification.id] is now set to the notificationId which was originally sent to the NotificationManager
+        * */
+        private val MIGRATION_1_2: Migration = object : Migration(1, 2) {
+            override fun migrate(database: SupportSQLiteDatabase) {
+                database.execSQL("CREATE TABLE IF NOT EXISTS `notification_new` " +
```
Comment (darshanpania) [inline app/src/main/java/com/darshan/notificity/AppDatabase.kt:45]: @zzjjaayy Can we not create a new table and update the schema for the existing Notification table and update the primary keys?

File: app/src/main/java/com/darshan/notificity/NotificityListener.kt [hunk 3]
```diff
@@ -5,6 +5,7 @@ import android.content.pm.ApplicationInfo
 import android.content.pm.PackageManager.NameNotFoundException
 import android.service.notification.NotificationListenerService
 import android.service.notification.StatusBarNotification
+import android.util.Log
```
Comment (darshanpania) [inline app/src/main/java/com/darshan/notificity/NotificityListener.kt:8]: @zzjjaayy Log is not used anywhere

File: app/src/main/java/com/darshan/notificity/Notification.kt (hunk 1 above)
Comment (zzjjaayy) [reply app/src/main/java/com/darshan/notificity/Notification.kt]: Makes sense. Changed back.

File: app/src/main/java/com/darshan/notificity/AppDatabase.kt (hunk 2 above)
Comment (zzjjaayy) [reply app/src/main/java/com/darshan/notificity/AppDatabase.kt:45]: @darshanpania I think we cannot, sqlite doesn't let you change the primary keys like that. I did try and could not get it to work.

File: app/src/main/java/com/darshan/notificity/NotificityListener.kt (hunk 3 above)
Comment (zzjjaayy) [reply app/src/main/java/com/darshan/notificity/NotificityListener.kt]: correct. removed.

File: app/src/main/java/com/darshan/notificity/AppDatabase.kt (hunk 2 above)
Comment (mrwhoknows55) [reply app/src/main/java/com/darshan/notificity/AppDatabase.kt:45]: Checkout this https://stackoverflow.com/questions/16900552/change-the-primary-key-of-a-table-in-sqlite

Comment (darshanpania) [reply app/src/main/java/com/darshan/notificity/AppDatabase.kt:45]: thanks @zzjjaayy and @mrwhoknows55 for the info!

**Review by darshanpania — APPROVED** [review]
Approving the changes. Will merge and do a round of testing. Thank you @zzjjaayy for this PR. It fixes a very important problem I should not have introduced in the first place. 🙏

**Pull request merged**